src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.DeclarationAnalysisData.cs

```diff
@@ -4,51 +4,55 @@
+using Microsoft.CodeAnalysis.PooledObjects;
 
 namespace Microsoft.CodeAnalysis.Diagnostics
 {
     internal abstract partial class AnalyzerDriver
     {
-        internal sealed class DeclarationAnalysisData
+        internal readonly struct DeclarationAnalysisData : IDisposable
         {
-            public DeclarationAnalysisData(
-                SyntaxNode declaringReferenceSyntax,
-                SyntaxNode topmostNodeForAnalysis,
-                ImmutableArray<DeclarationInfo> declarationsInNodeBuilder,
-                ImmutableArray<SyntaxNode> descendantNodesToAnalyze,
-                bool isPartialAnalysis)
-            {
-                DeclaringReferenceSyntax = declaringReferenceSyntax;
-                TopmostNodeForAnalysis = topmostNodeForAnalysis;
-                DeclarationsInNode = declarationsInNodeBuilder;
-                DescendantNodesToAnalyze = descendantNodesToAnalyze;
-                IsPartialAnalysis = isPartialAnalysis;
-            }
-
             /// <summary>
             /// GetSyntax() for the given SyntaxReference.
             /// </summary>
-            public SyntaxNode DeclaringReferenceSyntax { get; }
+            public readonly SyntaxNode DeclaringReferenceSyntax;
 
             /// <summary>
             /// Topmost declaration node for analysis.
             /// </summary>
-            public SyntaxNode TopmostNodeForAnalysis { get; }
+            public readonly SyntaxNode TopmostNodeForAnalysis;
 
             /// <summary>
             /// All member declarations within the declaration.
             /// </summary>
-            public ImmutableArray<DeclarationInfo> DeclarationsInNode { get; }
+            public readonly ImmutableArray<DeclarationInfo> DeclarationsInNode;
 
             /// <summary>
             /// All descendant nodes for syntax node actions.
             /// </summary>
-            public ImmutableArray<SyntaxNode> DescendantNodesToAnalyze { get; }
+            public readonly ArrayBuilder<SyntaxNode> DescendantNodesToAnalyze = ArrayBuilder<SyntaxNode>.GetInstance();
 
             /// <summary>
             /// Flag indicating if this is a partial analysis.
             /// </summary>
-            public bool IsPartialAnalysis { get; }
+            public readonly bool IsPartialAnalysis;
+
+            public DeclarationAnalysisData(
+                SyntaxNode declaringReferenceSyntax,
+                SyntaxNode topmostNodeForAnalysis,
+                ImmutableArray<DeclarationInfo> declarationsInNodeBuilder,
+                bool isPartialAnalysis)
+            {
+                DeclaringReferenceSyntax = declaringReferenceSyntax;
+                TopmostNodeForAnalysis = topmostNodeForAnalysis;
+                DeclarationsInNode = declarationsInNodeBuilder;
+                IsPartialAnalysis = isPartialAnalysis;
+            }
+
+            public void Dispose()
+            {
+                DescendantNodesToAnalyze.Free();
```

Comment:
Should we do anything to defend against double frees? Maybe even something which is only done in DEBUG mode?

Reply:
`[NonCopying]` on containing type, plus changing this to `Nodes?.Free(); Nodes = null;`, might be adequate here.

Reply:
Also, genuinely don't know if I'm too far in the weeds bringing the concern up. I wasn't sure if this was something we'd already talked out and concluded on (non-idempotent Dispose in the compiler layer).

Reply:
Yeah, i'm really hesitant about how far to take this.  For example, we don't have people null'ing out their arraybuilders elsewhere after freeing them. If you do double-free, you should also get a debug assert in practically all scenarios.  So i think we're fine there.